### PR TITLE
add linenumber to JsonString

### PR DESCRIPTION
**Pull request by 928PJY**
add lineNumber to JsonString
@vwxyzh @hellosnow @superyyrrzz @qinezh @chenkennt

### CODE REVIEW

File: src/Microsoft.DocAsCode.MarkdownLite/JsonTokenTreeRenderer.cs [hunk 1]
```diff
@@ -42,7 +47,7 @@ public virtual StringBuffer Render(IMarkdownRenderer renderer, MarkdownBlockquot
             {
                 childContent += renderer.Render(item);
             }
-            return this.Insert("Blockquote", childContent, Type.NonLeaf);
+            return Insert(token, $"Blockquote", childContent);
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.MarkdownLite/JsonTokenTreeRenderer.cs]: >$ [](start = 33, length = 1)

remove

**Comment by vwxyzh** [comment]
:shipit:

File: src/Microsoft.DocAsCode.MarkdownLite/JsonTokenTreeRenderer.cs [hunk 2]
```diff
@@ -219,58 +224,68 @@ public virtual StringBuffer Render(IMarkdownRenderer renderer, MarkdownTextToken
...
-                        // TODO: add a method 'remove' of StringBuffer
-                        string contentTemp = tokenContent;
-                        tokenContent = contentTemp.Remove(contentTemp.Length - 1);
-                    }
-                    result += $",\"children\":[{tokenContent}]}},";
-                    break;
-                }
-                case Type.Leaf:
+                // If tokenContent is not empty ,should remove the last character(',')
+                if (tokenContent != StringBuffer.Empty && tokenContent.EndsWith(','))
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.MarkdownLite/JsonTokenTreeRenderer.cs]: >tokenContent != StringBuffer.Empty [](start = 20, length = 34)

can remove the check? EndsWith would check the length and return false, and line 248

File: src/Microsoft.DocAsCode.MarkdownLite/JsonTokenTreeRenderer.cs [hunk 3]
```diff
@@ -219,58 +224,68 @@ public virtual StringBuffer Render(IMarkdownRenderer renderer, MarkdownTextToken
...
+                // If tokenContent is not empty ,should remove the last character(',')
+                if (tokenContent.GetLength() > 0 && tokenContent.EndsWith(','))
+                {
+                    // TODO: add a method 'remove' of StringBuffer
+                    string contentTemp = tokenContent;
+                    tokenContent = contentTemp.Remove(contentTemp.Length - 1);
+                }
+                result += $",\"children\":[{tokenContent}]";
+            }
+            result += "},";
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.MarkdownLite/JsonTokenTreeRenderer.cs]: can reuse the overload method?

File: src/Microsoft.DocAsCode.MarkdownLite/JsonTokenTreeRenderer.cs [hunk 4]
```diff
@@ -121,12 +126,12 @@ public virtual StringBuffer Render(IMarkdownRenderer renderer, MarkdownTableBloc
                     {
                         childContent += renderer.Render(item);
                     }
-                    rowContent += this.Insert("RowItem", childContent, Type.NonLeaf);
+                    rowContent += Insert(row[j], "RowItem", childContent);
                 }
-                bodyContent += this.Insert("Row", rowContent, Type.NonLeaf);
+                bodyContent += Insert(row[0], "Row", rowContent);
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.MarkdownLite/JsonTokenTreeRenderer.cs]: >row[0] [](start = 38, length = 6)

is it row[0] or row [i]?

Comment (928PJY) [inline src/Microsoft.DocAsCode.MarkdownLite/JsonTokenTreeRenderer.cs]: it is row[0], for that row = Cells[i], so the row stand for the a row in the table,row[0] is the first child of a row,whose linenumber info is same with the this row.

---
In reply to: [77289159](https://github.com/dotnet/docfx/pull/663#discussion_r77289159) [](ancestors = 77289159)

**Comment by superyyrrzz** [comment]
:shipit:

File: src/Microsoft.DocAsCode.MarkdownLite/JsonTokenTreeRenderer.cs (hunk 4 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.MarkdownLite/JsonTokenTreeRenderer.cs]: got it.

---
In reply to: [77289954](https://github.com/dotnet/docfx/pull/663#discussion_r77289954) [](ancestors = 77289954,77289159)

**Pull request merged**